### ISSUE TITLE
fig.add_axes() with no arguments silently does nothing

### ISSUE THREAD

**Issue by marko-knoebl** [matplotlib/matplotlib#15059]
### Bug report

**Bug summary**

Running `fig.add_axes()` without any arguments does nothing (and doesn't raise an error).

**Code for reproduction**

```python
import matplotlib.pyplot as plt

fig = plt.figure()
new_axes = fig.add_axes()
print(new_axes)
```

**Actual outcome**

new_axes is None

the relevant code in add_axes reads:

```py
...
if not len(args):
    return
...
```

**Expected outcome**

either of these:

- raises an error
- returns an axis object with a default size (like pyplot.axes())

**Matplotlib version**

  * Matplotlib version: 3.0.3

**Comment by timhoffm** [comment]
IMHO this should behave the same as pyplot.axes().

**Comment by ImportanceOfBeingErnest** [comment]
So far, `fig.add_axes(rect)` returns an `Axes`. `plt.axes()` returns an `AxesSubplot`. To get a subplot, one should use `fig.add_subplot()`. I would not recommend letting `fig.add_axes()` return an `Axes` *or* an `AxesSubplot`, depending on the input. 

Hence, for consistency reasons, I would recommend to raise an error or warning.

**Comment by timhoffm** [comment]
Good point on the differences between `Axes` and `AxesSubplot`.

For now, we should raise a warning. Is there a reason we should not treat `add_axes()` equal to `add_axes([0, 0, 1, 1])`?

**Comment by ImportanceOfBeingErnest** [comment]
`add_axes([0, 0, 1, 1])` would create the axes as large as the figure, meaning axis decorators will be outside the figure, and users will be confused.  

One might be tempted to tread `add_axes()` equal to `add_axes([s.left, s.bottom, s.right-s.left, s.top-s.bottom])` (with `s = fig.subplotpars`), such that it looks exactly like the default subplot you get via `fig.add_subplot()`. However, since it is not an `AxesSubplot` it will not take part in layout management and upon calling `fig.tight_layout()` people would get a UserWarning. At this point it would be really hard to debug where that comes from.

So I think it should really raise a warning along the lines of 

> `UserWarning: Calling add_axes() without argument is unsupported. You may want to use add_subplot() instead.`

or deprecate it (with a similar warning) and let it raise an error in the future.

**Comment by timhoffm** [comment]
:+1: for deprecating and raising an error in the future.

**Comment by timhoffm** [comment]
Closed by #15286.
